src/services/completions.ts

```diff
@@ -535,14 +539,30 @@ namespace ts.Completions {
                 isJsxIdentifierExpected,
                 isRightOfOpenTag,
             );
-            getJSCompletionEntries(sourceFile, location.pos, uniqueNames, getEmitScriptTarget(compilerOptions), entries);
         }
-        else {
-            if (!isNewIdentifierLocation && (!symbols || symbols.length === 0) && keywordFilters === KeywordCompletionFilters.None) {
-                return undefined;
+
+        if (keywordFilters !== KeywordCompletionFilters.None) {
+            const entryNames = new Set(entries.map(e => e.name));
+            for (const keywordEntry of getKeywordCompletions(keywordFilters, !insideJsDocTagTypeExpression && isSourceFileJS(sourceFile))) {
+                if (isTypeOnlyLocation && isTypeKeyword(stringToToken(keywordEntry.name)!) || !entryNames.has(keywordEntry.name)) {
+                    insertSorted(entries, keywordEntry, compareCompletionEntries, /*allowDuplicates*/ true);
+                }
+            }
+        }
+
+        const entryNames = new Set(entries.map(e => e.name));
+        for (const keywordEntry of getContextualKeywords(contextToken, position)) {
+            if (!entryNames.has(keywordEntry.name)) {
+                insertSorted(entries, keywordEntry, compareCompletionEntries, /*allowDuplicates*/ true);
             }
+        }
 
-            getCompletionEntriesFromSymbols(
+        for (const literal of literals) {
+            insertSorted(entries, createCompletionEntryForLiteral(sourceFile, preferences, literal), compareCompletionEntries, /*allowDuplicates*/ true);
+        }
+
+        if (!isChecked) {
+            const uniqueNames = getCompletionEntriesFromSymbols(

```

Comment:
So this actually changes the order for unchecked files from

1. symbols
2. identifiers
3. keywords

to

1. keywords
2. symbols
3. identifiers

and I think we still want symbols first.

The (pre-existing) two huge calls to `getCompletionEntriesFromSymbols` are identical—can we lift and deduplicate that from the if/else? Then, let’s maintain `uniqueNames` from the result of that throughout the rest of this function so we don’t have to repeatedly regenerate it from `entries`.

Reply:
Ah, that makes sense. When I was moving the code around, I missed that `entries` was side-effected by `getCompletionEntriesFromSymbols`, and I assumed that the two calls were different since there would be no good reason to duplicate calls like that.

The code is much more elegant now, although reading it raises more questions than before. But those were questions that were hidden by the previous structure, in my opinion.

Reply:
Yeah, the important work is done by mutation and the return value is bonus info... not the best

Reply:
Yeah, the important work is done by mutation and the return value is bonus info... not the best